nemo/core/optim/mcore_optim.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+try:
+    from megatron.core.optimizer.optimizer import MegatronOptimizer
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    HAVE_MEGATRON_CORE = False
+
+
+class McoreDistributedOptimizer(torch.optim.Optimizer):
+    """
+    A wrapper for Mcore distributed optimizer.
+
+    Arguments:
+        optim: distributed optimizer from Megatron core.
+    """
+
+    def __init__(self, optim: MegatronOptimizer):
```

Comment:
You need to guard this type hint. You can do so by on line 25 setting
```
MegatronOptimizer = ApexGuardDefaults
```
See megatron_gpt_model.py for an example.

Reply:
Adding ApexGuardDefaults will have the circular import err. Looks like I should have a define my new Guard class in nemo.core ?
```
File "/lustre/fsw/coreai_dlalgo_llm/gaod/nemo/code/NeMo/nemo/core/optim/mcore_optim.py", line 16, in <module>
    from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults
  File "/lustre/fsw/coreai_dlalgo_llm/gaod/nemo/code/NeMo/nemo/collections/nlp/modules/__init__.py", line 16, in <module>
    from nemo.collections.nlp.modules.common import (
  File "/lustre/fsw/coreai_dlalgo_llm/gaod/nemo/code/NeMo/nemo/collections/nlp/modules/common/__init__.py", line 17, in <module>
    from nemo.collections.nlp.modules.common.bert_module import BertModule
  File "/lustre/fsw/coreai_dlalgo_llm/gaod/nemo/code/NeMo/nemo/collections/nlp/modules/common/bert_module.py", line 21, in <module>
    from nemo.core.classes import NeuralModule
ImportError: cannot import name 'NeuralModule' from partially initialized module 'nemo.core.classes' (most likely due to a circular import) (/lustre/fsw/coreai_dlalgo_llm/gaod/nemo/code/NeMo/nemo/core/classes/__init__.py)
[2024-04-28 22:39:38,287] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 1557884) of binary: /usr/bin/python
```

Reply:
Right now I just tried to remove two typing checks. Could you help trigger the test again? Thanks!